### PR TITLE
[6X] Refactor how we handle dropped column references in partition tables during major version upgrades

### PR DESCRIPTION
**Pull request by jimmyyih**
First commit:
```
Ignore partition table dropped columns for pg_dump --binary-upgrade

    When running pg_dump --binary-upgrade, dropped columns are always
    preserved to make sure the catalog matches with the physical data
    files. For GPDB partition tables, this gets very complicated because
    the root partition delegates whether or not the entire partition table
    will have dropped columns or not. If the root partition has dropped
    columns but a child partition does not, we hit a heterogeneous
    partition table scenario which requires manual user intervention. The
    current user workaround for this scenario is to do a plain CTAS on the
    entire partition table... but doing that could be extremely costly.

    To provide a less costly workaround, we simply need to ignore the
    dropped column from the partition table DDL that will be run on the
    target upgrade cluster and assume that all the child partitions will
    not have dropped columns (either it was already like that or the child
    partition has been fixed with a CTAS and exchange partition). However,
    if the root partition and all of its child partitions have the same
    dropped column reference, we will continue to output the dropped
    column in the schema dump as it is technically a homogeneous partition
    table and valid for upgrade. This solution works because the root
    partition does not store any data so now only the flagged child
    partitions (via pg_upgrade --check) need to be fixed instead of
    touching the entire partition table.

    Output format has also been minimalized to be similar to the Postgres
    check output (e.g. no "why" context). The context will be clarified
    from gpupgrade side.

    Note: This won't be a problem for GPDB 7+ because child partition DDL
    follows upstream Postgres logic where they are created separately and
    attached to the root/subroot.
```

Second commit:
```
Refactor pg_upgrade heterogeneous partition table check

    The heterogeneous partition table check would check if the attributes
    of a root partition and its child partitions were synchronized
    properly. If it detected an anomaly (e.g. root partition has dropped
    column references but a child partition does not), the partition table
    would be flagged and the user would be advised to recreate the table
    with CTAS. The workaround provided is correct but can be extremely
    costly in production clusters where partition tables can be very big.

    To lower the cost of fixing the issue, we now do much more complex
    queries and logic to determine which specific child partitions are
    problematic. We also now clearly define what heterogeneous/homogeneous
    partition tables are valid for upgrade. We then advise the user to
    CTAS only the problematic child partitions and do an exchange
    partition (the CTAS staging table would not have the dropped column
    reference). This workaround is a bit more complex and requires more
    concentrated user involvement but the time savings is pretty big
    (e.g. CTAS an entire partition table with 5000 child partitions
    vs. CTAS a couple child partitions).

    Note: This assumes that pg_dump --binary-upgrade will NOT output the
    dropped column reference in the partition table DDL schema dump if it
    detects a partition table where the root has dropped column references
    but its child partitions do not. Currently, there is a Greenplum hack
    to do specifically this.
```

Previous PR https://github.com/greenplum-db/gpdb/pull/13016 did not handle homogeneous partition tables where the root partition and all of its child partitions had the same dropped column reference (edge case regression).  This PR fixes that.

### CODE REVIEW

**Comment by kalensk** [comment]
Here is the PR for the associated pg_upgrade acceptance tests: https://github.com/greenplum-db/gpupgrade/pull/637 which is testing using this PR's branch to generate a test GPDB RPM for testing.

**Review by kalensk — APPROVED** [review]
Looks good from the pg_upgrade acceptance testing perspective!

**Review by soumyadeep2007 — COMMENTED** [review]
Some initial feedback on the first commit, Ill come back to this PR and re-review when I get more cycles.

File: src/test/regress/sql/pg_dump_binary_upgrade.sql [hunk 1]
```diff
@@ -0,0 +1,50 @@
+-- Ensure that pg_dump --binary-upgrade correctly suppresses the ALTER
```
[the remark this answers is not in the document]
Comment (jimmyyih) [reply src/test/regress/sql/pg_dump_binary_upgrade.sql:1]: Running 6X => 6X pg_upgrade is a common dev testing scenario IIRC.  But yes, gpupgrade 5X=>6X acceptance tests should "technically" cover this pg_dump scenario.  This test was added mainly to make sure the singular commit had its own test coverage within the repository.  If you think the pg_dump commit doesn't need a non-external test to cover it then that's fine with me.

[the remark this answers is not in the document]
Comment (kalensk) [reply src/test/regress/sql/pg_dump_binary_upgrade.sql:1]: (Just chiming in to say yes same version upgrades such as 6X->6X and 7X->7X are extremely common and extremely useful dev testing scenarios. We definitely need to be supporting those. In fact we have a story to fix an issue around this for 6->6 with tablespaces.)

File: src/test/regress/sql/pg_dump_binary_upgrade.sql [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+) DISTRIBUTED BY (b)
+PARTITION BY RANGE (a)
+(
+    PARTITION p1 START(1) END(5)
+);
+
+ALTER TABLE dump_this_schema.dropped_column_heterogeneous_partition_table DROP COLUMN c;
+CREATE TABLE dump_this_schema.staging_table_for_exchange_partition AS SELECT * FROM dump_this_schema.dropped_column_heterogeneous_partition_table_1_prt_p1;
+ALTER TABLE dump_this_schema.dropped_column_heterogeneous_partition_table EXCHANGE PARTITION p1 WITH TABLE dump_this_schema.staging_table_for_exchange_partition;
+DROP TABLE dump_this_schema.staging_table_for_exchange_partition;
```
[the remark this answers is not in the document]
Comment (jimmyyih) [reply src/test/regress/sql/pg_dump_binary_upgrade.sql:45]: Technically yes but I'd rather not pollute the pg_dump output with an uninteresting staging table.  The DROP TABLE is also a part of our workaround solution so its a bit of a documenting line as well.  Although if you want, I could remove it and update pg_dump to filter on the two partition tables instead (preferably not since the pg_dump line would become very long).  Thoughts?

**Comment by kalensk** [comment]
After discussing further offline, I think its best to output the two scenarios with headings as outlined below:
```
Database: isolation2test
  Partitions with Dropped Column References:
    public.differently_aligned_colu_1_prt_differently_alig_2_prt_subpart_1
    public.differently_aligned_varl_1_prt_differently_alig_2_prt_subpart_1
    public.differently_sized_column_1_prt_differently_size_2_prt_subpart_1

  Misaligned Dropped Column References:
    public.dropped_column_1_prt_part_2
```

This will allow us to easily document their respective workarounds and explanations.

File: src/test/regress/sql/pg_dump_binary_upgrade.sql (hunk 2 above)
[the remark this answers is not in the document]
Comment (soumyadeep2007) [reply src/test/regress/sql/pg_dump_binary_upgrade.sql:45]: > The DROP TABLE is also a part of our workaround solution so its a bit of a documenting line as well

Ah thats motivation enough to keep it.

File: src/test/regress/sql/pg_dump_binary_upgrade.sql (hunk 1 above)
[the remark this answers is not in the document]
Comment (soumyadeep2007) [reply src/test/regress/sql/pg_dump_binary_upgrade.sql:1]: Agreed, lets keep the test and the comment  is not necessary.

File: contrib/pg_upgrade/greenplum/check_gp.c [hunk 3]
```diff
@@ -497,102 +526,132 @@ check_heterogeneous_partition(void)
...
 		pg_log(PG_REPORT, "fatal\n");
 		pg_log(PG_FATAL,
-			   "| Your installation contains heterogenous partitioned tables\n"
-			   "| where the root partition does not match one or more child\n"
-			   "| partitions' on disk representation. In order to make the\n"
-			   "| tables homogenous, create a new partition table with the same\n"
-			   "| schema as the old partition table, insert the old data into\n"
-			   "| the new table, and drop the old table.\n"
+			   "| Your installation contains heterogeneous partitioned tables\n"
+			   "| where child partitions have dropped column references. Upgrade\n"
```
Comment (soumyadeep2007) [inline contrib/pg_upgrade/greenplum/check_gp.c:647]: `where certain child partitions have dropped column references that are not present in the root partition or other child partitions. Upgrade cannot reliably generate ALTER statements necessary to preserve the dropped columns for such partitions. This is because ALTER statements can only be applied from the root partition, which will affect all tables in the hierarchy. Preservation of these columns is vital for the on-disk compatibility of the child partition. Thus, upgrade cannot proceed.`

**Review by soumyadeep2007 — COMMENTED** [review]
Due to how tightly coupled the two commits are, for the sake of atomicity, we should squash them **(only before merging).** There are a number of places where the first commit would benefit from referring to the second and vice-versa.

Review todos:

1) Vet the 2nd commit a little more, especially the 2 queries. The first one seems good. Need to dive into the 2nd one.

2) Help come up w/ a unifying commit message for the squash.

PS: I have also provided some initial actionable feedback for the gpupgrade PR.

File: contrib/pg_upgrade/greenplum/check_gp.c (hunk 3 above)
Comment (jimmyyih) [reply contrib/pg_upgrade/greenplum/check_gp.c:647]: @kalensk What do you think about this suggested change?  I believe my previous iterations of this paragraph also had more detail but we simplified it together and deferred the specifics to the gpupgrade docs.

Comment (kalensk) [reply contrib/pg_upgrade/greenplum/check_gp.c:647]: I like @soumyadeep2007's sentence mentioning on-disk compatibility. I prefer to keep the DDL statement vague and not overloaded with specifics such as the ALTER. I prefer we explain and mention the two checks we have using the same or similar wording: `Partitions with Dropped Column References` and `Misaligned Dropped Column References`

Thus, my updated suggested text:

> Your installation contains heterogeneous partition tables. Either one or more child partitions have 1) dropped column references, or 2) are misaligned from their siblings in their name, type, length, or alignment. Upgrade cannot output partition table DDL to preserve the dropped columns for the detected child partitions. Preservation of these columns is necessary for on-disk compatibility. In order to correct the child partitions, create a new staging table with the same schema as the child partition, insert the old data into the staging table, exchange the child partition with the staging table, and drop the staging table. Alternatively, the entire partition table can be recreated instead.

Everyone will have their own take on the wording…

File: contrib/pg_upgrade/greenplum/check_gp.h [hunk 4]
```diff
@@ -6,18 +6,55 @@
...
+	"            JOIN pg_catalog.pg_namespace n2 ON n2.oid = c2.relnamespace " \
+	"            JOIN pg_catalog.pg_attribute a2 ON c2.oid = a2.attrelid AND a2.attnum > 0 " \
+	"        GROUP BY c2.oid, n2.nspname, c2.relname, c2.relnatts " \
+	"    ) t2 ON t2.oid = t1.parrelid " \
+	"WHERE NOT (t2.parnatts = t1.childnatts AND t2.parnumattisdropped = 0 AND t1.childnumattisdropped = 0) " \
+	"ORDER BY t2.parrelname, t1.childnatts, t1.childnumattisdropped;" \
 
+/*
+ * This query gathers information on root and child partition attributes that
+ * may possibly have mismatched names, dropped column references, types,
```
[the remark this answers is not in the document]
Comment (jimmyyih) [reply contrib/pg_upgrade/greenplum/check_gp.h:42]: The query compares attributes via their designated attribute number.  It is possible to have mismatched names such as the following example:
```
root partition
att1 - a
att2 - b
att3 - c
att4 - d
att5 - ...dropped...

child partition
att1 - a
att2 - b
att3 - ...dropped...
att4 - c
att5 - d
```
So for the above example, attributes 3 through 5 have name mismatches.  Technically, the other types of misalignment detection should always accompany but its always good to have peace of mind with a simple attname check.

File: contrib/pg_upgrade/greenplum/check_gp.c (hunk 3 above)
Comment (jimmyyih) [reply contrib/pg_upgrade/greenplum/check_gp.c]: Updated the paragraph with a combination of all the previous/suggested wordings.

File: contrib/pg_upgrade/greenplum/check_gp.c [hunk 5]
```diff
@@ -497,102 +526,132 @@ check_heterogeneous_partition(void)
...
+					 * 2. The child partition has the same number of dropped
+					 *    column references as the root partition, the number
+					 *    of columns match, and it has been determined that
+					 *    the child partitions should have dropped column
+					 *    references.
+					 */
+					if ((childnumattisdropped == 0) ||
+						(childnumattisdropped == parnumattisdropped && parnatts == childnatts &&
+						 expect_child_partitions_have_dropped_columns))
+						continue;
```
[the remark this answers is not in the document]
Comment (jimmyyih) [reply contrib/pg_upgrade/greenplum/check_gp.c]: I also thought that but went the safe route of checking both since I wasn't 100% certain.  Anyways, I've completely refactored this part and was able to create a query that detects all the problematic child partitions (which was kinda influenced by now assuming `childnumattisdropped == parnumattisdropped` implying `parnatts == childnatts`).  The C logic should be a lot simpler to follow now (and technically the query as well).

File: contrib/pg_upgrade/greenplum/check_gp.h [hunk 6]
```diff
@@ -6,18 +6,72 @@
...
+	"    JOIN pg_catalog.pg_attribute att1 ON att1.attrelid = par.parrelid " \
+	"    JOIN pg_catalog.pg_attribute att2 ON att2.attrelid = rule.parchildrelid AND att1.attnum = att2.attnum AND att1.attnum > 0 " \
+	"      AND NOT (att1.attisdropped = att2.attisdropped AND att1.attname = att2.attname AND att1.attlen = att2.attlen AND " \
+	"               att1.atttypid = att2.atttypid AND att1.attalign = att2.attalign) " \
+	"WHERE rule.parchildrelid IN ( " \
+	"    SELECT pr.parchildrelid " \
+	"    FROM pg_catalog.pg_partition_rule pr " \
+	"        JOIN pg_catalog.pg_class c ON c.oid = pr.parchildrelid AND NOT c.relhassubclass " \
+	"        JOIN pg_catalog.pg_attribute a ON a.attrelid = pr.parchildrelid " \
+	"    WHERE a.attisdropped);"
```
Comment (soumyadeep2007) [inline contrib/pg_upgrade/greenplum/check_gp.h:77]: Right now the query can return something like:
```
          parrelid           | attnum |              parchildrelid               |           attname1           |           attname2           | attisdropped1 | attisdropped2 | attlen1 | attlen2 | atttypid1 | atttypid2 | attalign1 | attalign2 
-----------------------------+--------+------------------------------------------+------------------------------+------------------------------+---------------+---------------+---------+---------+-----------+-----------+-----------+-----------
 different_dropped_col_order |      2 | different_dropped_col_order_1_prt_part_1 | b                            | ........pg.dropped.2........ | f             | t             |       4 |       4 |        23 |         0 | i         | i
 different_dropped_col_order |      3 | different_dropped_col_order_1_prt_part_1 | ........pg.dropped.3........ | b                            | t             | f             |       4 |       4 |         0 |        23 | i         | i
(2 rows)
```

The client code that calls this query iterates over the result set and prints out the partition name. In this case, we would end up printing `different_dropped_col_order_1_prt_part_1` twice.

So I would suggest replacing the entire projection list w/ `SELECT DISTINCT rule.parchildrelid::regclass`. That's all we need in the client code, so let's just project that.

File: contrib/pg_upgrade/greenplum/check_gp.h [hunk 7]
```diff
@@ -6,18 +6,72 @@
...
+ * This query gathers all the root (rp) and child (cp1 and cp2) partition
+ * information necessary to detect if a partition table is a heterogeneous
+ * partition table. The WHERE filter compares the root and child partition
+ * information to exclude out the valid homogeneous partition table
+ * definitions (as documented in the check_heterogeneous_partition() header).
+ *
+ * Note: We ignore subroot partitions because they also do not have data like
+ * the root partition.
+ */
+#define CHECK_PARTITION_TABLE_DROPPED_COLUMN_REFERENCES \
```
Comment (soumyadeep2007) [inline contrib/pg_upgrade/greenplum/check_gp.h:19]: I'm truly having difficulty in understanding this query (maybe its just me). Can you explain cp1 and cp2 a bit?

**Review by soumyadeep2007 — COMMENTED** [review]
Summary:
* Q2 looks good.
* Q1 and Q2 both pass all of the test cases except the ones mentioned inline.
* Both Q1 and Q2 can be minimized wrt projection list.
* Kalen and I saw output of the form:
```
  Partitions with invalid dropped column references:
  Partitions with misaligned columns:
    public.child_has_dropped_column_1_prt_part_1
    public.diff_no_dropped_columns_1_prt_part_1
    public.diff_no_dropped_columns_1_prt_part_2
    public.differently_aligned_colu_1_prt_differently_alig_2_prt_subpart_1
    public.differently_aligned_varl_1_prt_differently_alig_2_prt_subpart_1
    public.differently_sized_column_1_prt_differently_size_2_prt_subpart_1
    public.dropped_column_1_prt_part_2
```
I don't know how the above is possible. Reading the code suggests that this can't really happen. Maybe it is a bug in the test, rather than this PR.

**Comment by kalensk** [comment]
Slight nit: Let's not print out the headers if no results were detected for that check. That is, omit printing the header `Partitions with invalid dropped column references` or `Partitions with misaligned columns` if no results were found for them.

File: contrib/pg_upgrade/greenplum/check_gp.h (hunk 7 above)
Comment (jimmyyih) [reply contrib/pg_upgrade/greenplum/check_gp.h:19]: Yeah, I don't think it's you; SQL queries can be hard to understand depending on a lot of factors (e.g. is the author SQL proficient, is the reader SQL proficient, does the reader understand the reasoning behind the query, etc).

The cp1 subquery gets all the child partitions and some info about them: the number of attributes and how many dropped column references they have.  The cp2 subquery gets the min/max number of attributes amongst all the child partitions.  The two are then combined with a JOIN on the child partition's root partition oid.  I tried to use a window function but struggled for hours... and then was able to come up with the cp2 subquery instead.  The min/max child partitions natts is used in the complex case where the root partition has dropped column references but some child partitions do not.

File: contrib/pg_upgrade/greenplum/check_gp.h [hunk 8]
```diff
@@ -6,18 +6,72 @@
...
+	"SELECT rp.oid AS parrelid, rp.parnamespace, rp.parrelname, cp1.childnamespace, cp1.childrelname, " \
```
[the remark this answers is not in the document]
Comment (jimmyyih) [reply contrib/pg_upgrade/greenplum/check_gp.h:20]: Okay, so here's the SQL for this particular issue:
```
CREATE TABLE diff_no_dropped_columns (a int, b int, c char, to_drop varchar(50)) PARTITION BY RANGE (a) (PARTITION part_1 START(1) END(5), PARTITION part_2 START(5));
ALTER TABLE diff_no_dropped_columns DROP COLUMN to_drop;

CREATE TABLE intermediate_table (a int, b int, c char, to_drop varchar(50), to_drop_2 int);
ALTER TABLE intermediate_table DROP COLUMN to_drop;
ALTER TABLE intermediate_table DROP COLUMN to_drop_2;

ALTER TABLE diff_no_dropped_columns EXCHANGE PARTITION part_1 WITH TABLE intermediate_table;
DROP TABLE intermediate_table;
```

The attributes go like this in the catalog:
```
root(a, b, c, ...dropped1...)
child1(a, b, c, ...dropped1..., ...dropped2...)
child2(a, b, c, ...dropped1...)
```

There are two choices here which can be somewhat complex to think about:
1) Flagging all child partitions here is correct.
    a) If only child1 is fixed with the recommended workaround (automated by migration scripts), then child2 (and essentially all other child partitions if exists) will be flagged on the second pg_upgrade run.  If even one child partition does not have the same dropped column reference, then all other child partitions must go through the workaround which is similar to this above scenario.
    b) Technically, child1 can be fixed with a more complicated workaround (creating the staging table with ...dropped1... column) but this would make the new gpupgrade migration scripts extremely complicated for the `invalid dropped column reference` pg_upgrade check (the 1st query).  This would be similar to the manual workaround that users would need to make if the misaligned attribute check (the 2nd query) flags anything (will be part of gpupgrade documentation since it can't be easily automated and would probably require support assistance).
2. The partition table is actually completely upgradeable assuming the ...dropped2... attribute in child1 doesn't matter.
    a) If this is correct and we want to exclude this from the heterogeneous partition check, then we would have to do a pretty big refactor to the pg_dump check, and both pg_upgrade checks to add this extreme edge-case.

**Comment by jimmyyih** [comment]
>     Kalen and I saw output of the form:
>
>  Partitions with invalid dropped column references:
  Partitions with misaligned columns:
    public.child_has_dropped_column_1_prt_part_1
    public.diff_no_dropped_columns_1_prt_part_1
    public.diff_no_dropped_columns_1_prt_part_2
    public.differently_aligned_colu_1_prt_differently_alig_2_prt_subpart_1
    public.differently_aligned_varl_1_prt_differently_alig_2_prt_subpart_1
    public.differently_sized_column_1_prt_differently_size_2_prt_subpart_1
    public.dropped_column_1_prt_part_2
>
> I don't know how the above is possible. Reading the code suggests that this can't really happen. Maybe it is a bug in the test, rather than this PR.


> Slight nit: Let's not print out the headers if no results were detected for that check. That is, omit printing the header Partitions with invalid dropped column references or Partitions with misaligned columns if no results were found for them.

These were found to be issues in the tests rather than this PR because the tests sort the output.  I did update the ORDER BY in the 1st query which should help make that deterministic but the 2nd query felt weird to add an ORDER BY (would have to add the ORDER BY attributes to the DISTINCT as well because DISTINCT is technically an aggregate).  FWIW, I ran the pg_upgrade check 1000 times and the output file always had the same line ordering for its contents.

**Comment by soumyadeep2007** [comment]
> These were found to be issues in the tests rather than this PR because the tests sort the output. I did update the ORDER BY in the 1st query which should help make that deterministic but the 2nd query felt weird to add an ORDER BY (would have to add the ORDER BY attributes to the DISTINCT as well because DISTINCT is technically an aggregate). FWIW, I ran the pg_upgrade check 1000 times and the output file always had the same line ordering for its contents.

Ordering the results from these checks feels weird for the sake of test diffing. Just another thought, could we break these up into 2 separate pg_upgrade checks? These need separate documentation, workarounds anyway and it would make it easier on the gpupgrade tests as well.

File: contrib/pg_upgrade/greenplum/check_gp.h (hunk 8 above)
[the remark this answers is not in the document]
Comment (soumyadeep2007) [reply contrib/pg_upgrade/greenplum/check_gp.h]: I don't like the idea of 2 (what if you have `child1(a, ...dropped1..., ...dropped2..., b, c)` ?

Throwing this out there, is there a way for Q1 to ignore the case you mentioned above and for Q2 to flag it?
(I tried Q2 directly on the case, but it didn't flag anything)

[the remark this answers is not in the document]
Comment (jimmyyih) [reply contrib/pg_upgrade/greenplum/check_gp.h]: > I don't like the idea of 2 (what if you have child1(a, ...dropped1..., ...dropped2..., b, c) ?

Then the second query would flag it for manual fixing (see 1b).

> Throwing this out there, is there a way for Q1 to ignore the case you mentioned above and for Q2 to flag it?
(I tried Q2 directly on the case, but it didn't flag anything)

I tried for a couple hours to do just that.  Let me see if a couple more hours can produce anything.

**Comment by jimmyyih** [comment]
> Ordering the results from these checks feels weird for the sake of test diffing.

The ORDER BY was originally there for the old complicated C logic but I kept it to make it better for users to have the problematic child partitions be grouped to the root partition they belong to (would be kinda annoying if they were randomly mixed around).

> Just another thought, could we break these up into 2 separate pg_upgrade checks? These need separate documentation, workarounds anyway and it would make it easier on the gpupgrade tests as well. Added benefit is that the test files in the gpupgrade PR would be reduced in size as well.

Yes, that is doable if we want to further refactor the heterogeneous partition table check.  Originally, I had thought the same but restrained from doing so because I didn't think initially that the changes would be somewhat complex.  If that's how we want to proceed then sure.

**Comment by kalensk** [comment]
@orhankislal helped figure out how to effectively split on the file and sort. That is, we split the file and sort the two sub-checks individually. So I think we are good! Thanks @orhankislal!

File: contrib/pg_upgrade/greenplum/check_gp.h [hunk 9]
```diff
@@ -47,32 +51,44 @@ void check_heterogeneous_partition(void);
...
  * ...dropped...) and child(a, ...dropped..., b). In the query below, we
- * exclude partition tables where only the root partition has dropped column
+ * exclude partition tables where the root partition does not have dropped
+ * column references or only the root partition has dropped column
  * references. We also exclude subroot partitions because they also do not
  * have data like the root partition.
  */
-#define CHECK_PARTITION_TABLE_MATCHES_COLUMN_ATTRIBUTES \
-	"SELECT DISTINCT rule.parchildrelid::regclass " \
+#define CHECK_PARTITION_TABLE_MATCHES_DROPPED_COLUMN_ATTRIBUTES \
```
Comment (soumyadeep2007) [inline contrib/pg_upgrade/greenplum/check_gp.h:74]: As discussed, I slightly rewrote the query and ran it against cases 3 to 7 in the gpupgrade PR:

```
WITH root_dropped_attr AS
(SELECT par.oid AS paroid, *
    FROM pg_catalog.pg_partition par JOIN pg_catalog.pg_attribute a
    ON par.parrelid = a.attrelid
    WHERE NOT par.paristemplate AND a.attisdropped)
SELECT DISTINCT child_dropped_attr.parchildrelid::regclass
FROM root_dropped_attr
    RIGHT JOIN (
        SELECT pr.paroid, pr.parchildrelid, a.attnum, a.attisdropped, a.attname, a.attlen, a.atttypid, a.attalign
        FROM pg_catalog.pg_partition_rule pr
            JOIN pg_catalog.pg_class c ON c.oid = pr.parchildrelid AND NOT c.relhassubclass
            JOIN pg_catalog.pg_attribute a ON a.attrelid = pr.parchildrelid
        WHERE a.attisdropped AND pr.paroid IN (SELECT DISTINCT paroid FROM root_dropped_attr)
    ) child_dropped_attr
    ON root_dropped_attr.paroid = child_dropped_attr.paroid
        AND root_dropped_attr.attnum = child_dropped_attr.attnum
WHERE root_dropped_attr.attisdropped IS DISTINCT FROM child_dropped_attr.attisdropped
      OR root_dropped_attr.attname IS DISTINCT FROM child_dropped_attr.attname
      OR root_dropped_attr.attlen IS DISTINCT FROM child_dropped_attr.attlen
      OR root_dropped_attr.atttypid IS DISTINCT FROM child_dropped_attr.atttypid
      OR root_dropped_attr.attalign IS DISTINCT FROM child_dropped_attr.attalign;
```

It gave the same results.

Things I did:
* Used a CTE for the root tuple
* Renamed the tuple variables
* Pushed paristemplate and attisdropped into the individual child and root subqueries in an attempt to have the JOIN ON clause only consist of equality condition(s) that affect what gets joined.

Comment (soumyadeep2007) [reply contrib/pg_upgrade/greenplum/check_gp.h:74]: Top-level cost comparison (seems like the CTE pays off):
Original : ` HashAggregate  (cost=56.84..56.92 rows=8 width=4)`
Rewritten: ` HashAggregate  (cost=152.68..152.69 rows=1 width=4)`

Comment (jimmyyih) [reply contrib/pg_upgrade/greenplum/check_gp.h:74]: Great, so the CTE worked on 5X_STABLE.  I've slightly formatted the suggested CTE query and replaced the previous 2nd query with it.  Manually tested it with my own sanity DDL as well.

**Comment by jimmyyih** [comment]
@soumyadeep2007 All feedback addressed and manually tested.

> Due to how tightly coupled the two commits are, for the sake of atomicity, we should squash them (only before merging). There are a number of places where the first commit would benefit from referring to the second and vice-versa.

Do you still think the two commits should be squashed?  I can see them being separate but I understand the benefits of squashing them as well since the special logic are kinda tied together between pg_upgrade and pg_dump.  If so, I'll squash and redo the commit message before merging.

**Review by soumyadeep2007 — APPROVED** [review]
LGTM. I vote for squashing it into one commit due to the coupling.
Squashing would give us an opportunity to directly refer to the code in pg_dump, in order to improve comments like:
```
* If anything is to change here, please review if anything needs to be
 * changed in the related GPDB hack for pg_dump --binary-upgrade.
```
